### PR TITLE
Schema intersection

### PR DESCRIPTION
**Pull request by eddyxu**
* Refactor `datatype` module to separate files.
* Will be used by `Fragment` to decide which data file contains the columns to read.

### CODE REVIEW

File: rust/src/datatypes/schema.rs [hunk 1]
```diff
@@ -0,0 +1,475 @@
...
+        }
+
+        Ok(Self {
+            fields: candidates,
+            metadata: self.metadata.clone(),
+        })
+    }
+
+    /// Intersection between two [`Schema`].
+    pub fn intersection(&self, other: &Self) -> Result<Self> {
```
Comment (eddyxu) [inline rust/src/datatypes/schema.rs:72]: This is what is added in this PR.

Comment (changhiskhan) [reply rust/src/datatypes/schema.rs:72]: so intersection only happens at the top level then? do we need to also look at child fields?

**Review by changhiskhan — APPROVED** [review]

File: rust/src/datatypes/schema.rs (hunk 1 above)
Comment (eddyxu) [reply rust/src/datatypes/schema.rs:72]: Yes. Only these.

File: rust/src/datatypes/field.rs [hunk 2]
```diff
@@ -0,0 +1,648 @@
...
+    encodings::Encoding,
+    format::pb,
+    io::object_reader::{read_binary_array, read_fixed_stride_array, ObjectReader},
+    Error, Result,
+};
+
+/// Lance Schema Field
+///
+#[derive(Debug, Clone, PartialEq)]
+pub struct Field {
```
[the remark this answers is not in the document]
Comment (eddyxu) [reply rust/src/datatypes/field.rs:41]: it just refactor moves.

**Comment by eddyxu** [comment]
> so intersection only happens at the top level then? do we need to also look at child fields?


it does nested call into `Field::intersection()` which handles the nested structs.

**Pull request merged**